api/src/OpenTelemetry/Logs/Core.hs

```diff
@@ -124,41 +126,56 @@ setGlobalLoggerProvider :: (MonadIO m) => LoggerProvider body -> m ()
 setGlobalLoggerProvider = liftIO . writeIORef globalLoggerProvider
 
 
+defaultShutdownTimeout :: Int
+defaultShutdownTimeout = 5_000_000
+
+
 {- | This method provides a way for provider to do any cleanup required.
 
  This will also trigger shutdowns on all internal processors.
 -}
-shutdownLoggerProvider :: (MonadIO m) => LoggerProvider body -> m ()
-shutdownLoggerProvider LoggerProvider {loggerProviderProcessors} = liftIO $ do
-  asyncShutdownResults <- V.forM loggerProviderProcessors $ \processor -> do
-    logRecordProcessorShutdown processor
-  mapM_ wait asyncShutdownResults
+shutdownLoggerProvider
+  :: (MonadIO m)
+  => Maybe Int
+  -- ^ Optional timeout in microseconds, defaults to 5,000,000 (5s)
+  -> LoggerProvider body
+  -> m ShutdownResult
+  -- ^ Result that denotes whether the shutdown action succeeded, failed, or timed out.
+shutdownLoggerProvider mtimeout LoggerProvider {loggerProviderProcessors} = liftIO $ do
+  mresult <-
+    timeout (fromMaybe defaultShutdownTimeout mtimeout) $
+      takeWorstShutdownResult
+        <$> forConcurrently
+          loggerProviderProcessors
+          (\lrp -> logRecordProcessorShutdown lrp `catch` \(SomeException _) -> pure ShutdownFailure)
```

Comment:
I'm not sure where the best place is to catch these errors, but they must be caught so they do not affect programs using instrumentation. I'm not sure if catching `SomeException` is the best strategy. [This is relevant](https://opentelemetry.io/docs/specs/otel/error-handling/#guidance), especially point number 4.